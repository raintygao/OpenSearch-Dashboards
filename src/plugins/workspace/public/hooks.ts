/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { useObservable } from 'react-use';
import { useMemo } from 'react';
import { of } from 'rxjs';
import { WorkspaceFeature, WorkspaceFeatureGroup } from './components/workspace_form/types';
import { ApplicationStart, PublicAppInfo } from '../../../core/public';
import { isWorkspaceFeatureGroup } from './components/workspace_form/utils';
import { DEFAULT_SELECTED_FEATURES_IDS } from '../common/constants';

export function useApplications(application?: ApplicationStart) {
  const applications = useObservable(application?.applications$ ?? of(new Map()), new Map());
  return useMemo(() => {
    const apps: PublicAppInfo[] = [];
    applications?.forEach((app) => {
      apps.push(app);
    });
    return apps;
  }, [applications]);
}

export function useAllWorkspaceFeatures(
  workspaceFeatureOrGroup: Array<WorkspaceFeature | WorkspaceFeatureGroup>
) {
  const registerFeatureIds = workspaceFeatureOrGroup.reduce<string[]>((acc, cur) => {
    if (isWorkspaceFeatureGroup(cur)) {
      const ids = cur.features.map((feature) => {
        return feature.id;
      });
      return acc.concat(ids);
    } else {
      acc.push(cur.id);
      return acc;
    }
  }, []);
  return registerFeatureIds.concat(DEFAULT_SELECTED_FEATURES_IDS);
}
