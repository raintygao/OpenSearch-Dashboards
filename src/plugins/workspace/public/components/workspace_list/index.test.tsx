/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { shallow } from 'enzyme';
import { WorkspaceList } from './index';
import { coreMock } from '../../../../../core/public/mocks';
import { render, waitFor } from '@testing-library/react';

import { of } from 'rxjs';

jest.doMock('../../../../opensearch_dashboards_react/public', () => ({
  useOpenSearchDashboards: jest.fn().mockReturnValue({
    services: coreMock.createStart(),
  }),
}));

test('render normally', () => {
  const component = shallow(<WorkspaceList />);
  expect(component).toMatchSnapshot();
});

describe('WorkspaceList', function () {
  it('should render title and table regularly', () => {
    const { getByText, getByRole } = render(<WorkspaceList />);
    expect(getByText('Workspaces')).toBeInTheDocument();
    expect(getByRole('table')).toBeInTheDocument();
  });

  it('should render data in table based on workspace list', async () => {
    jest.doMock('../../../../opensearch_dashboards_react/public', () => ({
      useOpenSearchDashboards: jest.fn().mockReturnValue({
        services: {
          ...coreMock.createStart(),
          workspaces: {
            workspaceList$: of([
              { id: 'id1', name: 'name1' },
              { id: 'id2', name: 'name2' },
            ]),
          },
        },
      }),
    }));
    const { getByText } = render(<WorkspaceList />);
    await waitFor(() => {
      expect(getByText('name1')).toBeInTheDocument();
    });
  });
});
