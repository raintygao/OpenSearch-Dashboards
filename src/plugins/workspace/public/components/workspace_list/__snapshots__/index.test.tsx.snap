exports[`render 1`] = `
<EuiPage
  paddingSize="none"
>
  <EuiPageBody
    panelled={true}
  >
    <EuiPageHeader
      description="Workspace allow you to save and organize library items, such as index patterns, visualizations, dashboards, saved searches, and share them with other OpenSearch Dashboards users. You can control which features are visible in each workspace, and which users and groups have read and write access to the library items in the workspace."
      pageTitle="Workspaces"
      restrictWidth={true}
    />
    <EuiPageContent
      hasShadow={false}
      horizontalPosition="center"
      paddingSize="none"
      panelPaddingSize="l"
      style={
        Object {
          "maxWidth": 1000,
          "width": "100%",
        }
      }
      verticalPosition="center"
    >
      <EuiInMemoryTable
        columns={
          Array [
            Object {
              "field": "name",
              "name": "Name",
              "render": [Function],
              "sortable": true,
            },
            Object {
              "field": "id",
              "name": "ID",
              "sortable": true,
            },
            Object {
              "field": "description",
              "name": "Description",
              "truncateText": true,
            },
            Object {
              "field": "features",
              "hasActions": true,
              "isExpander": true,
              "name": "Features",
            },
            Object {
              "actions": Array [
                Object {
                  "description": "edit workspace",
                  "icon": "pencil",
                  "name": "Edit",
                  "onClick": [Function],
                  "type": "icon",
                },
              ],
              "field": "",
              "name": "Actions",
            },
          ]
        }
        isSelectable={true}
        itemId="id"
        items={Array []}
        onTableChange={[Function]}
        pagination={
          Object {
            "pageIndex": 0,
            "pageSize": 5,
            "pageSizeOptions": Array [
              5,
              10,
              20,
            ],
          }
        }
        responsive={true}
        search={
          Object {
            "box": Object {
              "incremental": true,
            },
            "onChange": [Function],
            "toolsRight": Array [
              <EuiButton
                href=""
              >
                Create workspace
              </EuiButton>,
            ],
          }
        }
        selection={
          Object {
            "onSelectionChange": [Function],
            "selectable": [Function],
          }
        }
        sorting={
          Object {
            "sort": Object {
              "direction": "asc",
              "field": "name",
            },
          }
        }
        tableLayout="fixed"
      />
    </EuiPageContent>
  </EuiPageBody>
</EuiPage>
`;
